add observer in `componentWillMount` in nuclear component decorator

dtothefp opened the pull request:
@Sinewyk I updated my sample repo to demonstrate an issue I'm having with the `nuclearComponent` decorator https://github.com/dtothefp/nuclear-decorator/blob/master/app/components/modal/index.jsx.

This is a rough example of a modal with two cases
- opens from a user action such as clicking a button
- opens from an environmental state, such as cookie or query parameter.

The first situation works as expected, where as the state/component gets out of sync when doing the environmental state logic check in my modal component in `componentWillMount` or `componentDidMount`. Essentially from this SO post http://stackoverflow.com/questions/23734862/check-if-all-child-components-have-been-mounted/23744270#23744270 where you are attaching the change observer will not be called until after the lifecycle methods I want to hook into for initial environment checks have been fired. When `component.context.reactor.observe` is moved inside of `componentWillMount` for the decorator component, my modal works as expected.

In addition @colindresj proposes this change because "another reason for moving setState out of componentDidMount for the decorator is that did mount is not called on the server, making nuclear component not client-server universal"

Sinewyk commented:
Could you be more precise for what is not working in your example ?

You probably need to push those side effects (environmental checks into components ? => that component is not universal) outside your components. Call reactor.observe in your entry files either client or server side.

dtothefp commented:
- clone the example repo
- run `npm start` and open the console in the browser window that opens
- refresh the page because the first time you land it will try to open the modal if a cookie isn't present and the syncing problem I'm speaking of will have already happened. 
- hit the button to open the modal, watch the console output, see that the prop for `isOpen` is being toggled to hide/show the modal
- add the query param `?p=t` to the url and hit enter to refresh the page
- the modal is supposed to open if this parameter is not present, but with the current code of the decorator in combination with the component it won't. Hit the open modal button a few times and see that the component and decorator will not re-render because `reactor.observe` was never called in the decorator `componentDidMount`.
- now change the code in the decorator to what my PR changes and see that the modal will open and close properly as expected. 

The environmental check is in the modal component because that is where it belongs, associated with the UI that it effects. I don't think it is a good solution to move it into another component, and I also don't think that will solve the problem with your current code because it will still be a child component of the React element created in the decorator http://stackoverflow.com/questions/23734862/check-if-all-child-components-have-been-mounted/23744270#23744270

Sinewyk commented:
I don't know where to start so I'll just dump some info in hope that you get my point :smile: 

The issue comes from how you are constructing your application. You are introducing problems by having a non pure component executing a synchronous action in the middle of your tree in it's render cycle.

Either it's component state (this.setState), lost when the component is unmounted, either it's application state (stored in the reactor) and should not be handled by your component lifecycle (or at least not synchronously, if you implement your data fetching in componentDidMount).

This is predictable application state, it should not live in your component, because it has actually nothing to do with your component.

Move your cookie/queryparser logic into your entry file (you are kickstarting your application state for the client side) then render.

You are now one step closer to universal-ing your app. You would also kickstart by instead sending req.url or reading the request cookies before rendering.

I'm going to send a small PR to your repo to show my case.

edit: example PR https://github.com/dtothefp/nuclear-decorator/pull/1

Sinewyk commented:
Also on a sidenote, you should never require your reactor from anywhere else that your entry file. Otherwise you are not universal ready, dunno if that was for speed or not.

https://github.com/optimizely/nuclear-js/blob/master/examples/isomorphic-flux-chat/modules/chat/actions.js#L1, that way server side when creating one reactor per request, you can kickstart stuff (locale, querystring if you want, user via cookie, etc etc) and then pass it through to react for rendering.

dtothefp commented:
hey @Sinewyk thanks for the feedback, we can probably close this PR and continue this discussion elsewhere.

Based upon your recommendations I've refactored off or your branch to bootstrap modal initial state from the `app.jsx`. Moved the only `import` of the reactor into this file as well.

https://github.com/dtothefp/nuclear-decorator/compare/dfp/track-modal-id

Also, added the ability to have multiple modals on the page that open based upon query params 

`http://localhost:8000/?home-lightbox=true&another-modal=true`

Let me know if you think this is a more solid architecture. I like it because state is set before the render cycle even begins so there is no lag in showing the modal if the state requirements are met

Sinewyk commented:
I like to think it is.

Server side you could do something similar to pre-render the modal by analyzing the request object before calling render.

Leaving this open for a bit, I'll close it later maybe when I've updated the readme to orient towards a better architecture maybe.

dtothefp commented:
@Sinewyk yeh a little more overview of proper architecture and difference in implementation of decorators vs. mixins would be helpful.

Sinewyk commented:
https://github.com/jordangarcia/nuclear-js-react-addons/pull/6

dtothefp commented:
I'd like to continue this thread with a different use case because I keep running into the same issue with the rendering lifecycle in relation to the `reactor.observe` registration in `componentDidMount` in `nuclearComponent`.

My current problem is a form component with children that I would like to register their initial data upon initial component lifecycle.

ex:

``` js
// decorator to provide additional context via props
import React, {Component, PropTypes, createClass} from 'react';
import {assign} from 'lodash';
import hoistNonReactStatics from 'hoist-non-react-statics';

function createComponent(Component, dataBindings) {
  const displayName = `ProvideContextComponent(${Component.displayName || Component.name})`;

  let ProvideContextComponent = createClass({
    displayName,

    contextTypes: dataBindings,

    render() {
      return React.createElement(Component, assign({}, this.props, this.state, this.context));
    }
  });

  hoistNonReactStatics(ProvideContextComponent, Component);

  return ProvideContextComponent;
}

export default function provideContextDec(Component, dataBindings) {
  if (arguments.length === 0 || typeof arguments[0] !== 'function') {
    dataBindings = arguments[0];
    return function connectToData(ComponentToDecorate) {
      return createComponent(ComponentToDecorate, dataBindings);
    };
  }

  return createComponent.apply(null, arguments);
}

//Form Component 
@nuclearComponent({
  formHasErrors: Getters.formHasErrors,
  inputState: Getters.validationState  //returns all state from the store
})
@provideContextDec({
  Actions: PropTypes.object.isRequired,
  formId: PropTypes.string.isRequired
})
export default class Form extends Component {
  constructor(props) {
    super(props);
    let {formId, Actions} = props;
    Actions.register(formId); //register the formId in the store, input data registers subsequently
  }

  handleSubmit(e) {
    //validation/submit logic here with dat from the store
  }

  render() {
    let classes = cx({
      'form-error': this.props.formHasErrors
    });

    return (
      <form onSubmit={::this.handleSubmit} className={classes}>
        {this.props.children}
        <button type="submit">Submit</button>
      </form>
    );
  }
}

//Input Component
@nuclearComponent({
  formErrors: Getters.formErrors,
  inputState: Getters.validationState
})
@provideContextDec({
  Actions: PropTypes.object.isRequired,
  formId: PropTypes.string.isRequired
})
export default class Input extends Component {
  constructor(props) {
    super(props);
    let {formId, Actions, defaultChecked, defaultValue} = props;
    if(defaultValue || defaultChecked) {
      Actions.registerFormField(formId, props); //register initial form values
    }
  }

  handleChange(e) {
    let {type, Actions, formId} = this.props;
    if(type === 'checkbox' || type === 'radio') {
      Actions.setFieldValue(formId, _.assign({}, this.props));
    } else {
      let {value} = e.target;
      Actions.setFieldValue(formId, _.assign({}, this.props, {value}));
    }
  }

  render() {
    let {type, name, ...props} = this.props;
    let activeInput = this.props.inputState.getIn([this.props.formId, 'currentValues', this.props.name])};
    let isActiveInput = activeInput.get('type') === type;
    let errors = this.props.formErrors && this.props.formErrors.getIn([this.formId, this.props.name]);
    let errorMsg = errors && errors.size && errors.get('error');
    let classes = cx(assign({}, makeClassesFromProps(this.props.classes), {
      error: errors && errors.size
    }));

   return (
      <input
        id={this.id}
        type={type}
        name={name}
        className={classes}
        onChange={::this.handleChange}
        value={isActiveInput ? activeInput.get('value') : null} //this won't work on initial render
        {...props}
      />
    );
  }
}

//Make the form
class SomeForm extends Comonent {
  render() {
    <Form>
      <Input type="text" name="text_input" defaultValue="eat" placeholder="burritos" />
      <Input type="radio" name="radio_input" defaultChecked={true} value="cool" />
    </Form>
  }
}

//Render the component
React.render(
  <SomeForm reactor={reactor} Actions={Actions} formId={_.uniqueId('form_')} />,
  document.body
);
```

The problem is with the render cycle in relation to the decorator, the order is:
1. Getters are registered inside of the decorator // not sure if this is the correct order?
2. The form component registers it's id
3. All inputs register their form values
4. componentWillMount, render, and componentDidMount fire for the input components
5. componentWillMount, render, and componentDidMount fire for the form component
6. render/componentDidMount fires for the decorator and the `reactor.observe` is established

Therefore, the input props such as `value={isActiveInput ? activeInput.get('value') : null}` in the `render` function for inputs is trying to read undefined data from the props because the `Getter`s have not yet been evaluated. Again if I change the `nuclearComponent` call of `reactor.observe` to `componentWillMount` the problem is solved. 

For now I've been hacking through it by iterating through all form chidren and gathering initial data using `this.props.children` but this is less than ideal. Any advice you could offer on where/how to establish gathering initial form data so that data sent to the stores will be available in the initial `render` of the input/form components would be extremely useful :neckbeard:

Sinewyk commented:
You probably should reverse where the data is coming in. Actions should probably only be triggered by a user's action, or by a flow from the application to fetch data from the server not by some weird initialize process sparkled among the many inputs.

Your `SomeForm` could probably kickstart the initial value of your form in its constructor all at once.

And again, I would go farther and move all of the state out of the component and move it in nuclear by having a store for this form with an initial state of what I want and simply render it, same for the errors.

dtothefp commented:
@Sinewyk again you're right. Initializing input data outside of components by registering data in a store before `React.render` cleaned up the logic/functionality alot

Sinewyk commented:
No problem ^^, happy to help.
